src/PhpSpreadsheet/Calculation/LookupRef/Helpers.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace PhpOffice\PhpSpreadsheet\Calculation\LookupRef;
+
+use PhpOffice\PhpSpreadsheet\Cell\AddressHelper;
+use PhpOffice\PhpSpreadsheet\Cell\Cell;
+use PhpOffice\PhpSpreadsheet\Spreadsheet;
+use PhpOffice\PhpSpreadsheet\Worksheet\Worksheet;
+
+class Helpers
+{
+    private static function convertR1C1(string &$cellAddress1, ?string &$cellAddress2, bool $a1): string
+    {
+        if (!$a1) {
+            $cellAddress1 = AddressHelper::convertToA1($cellAddress1);
+            if ($cellAddress2) {
+                $cellAddress2 = AddressHelper::convertToA1($cellAddress2);
+            }
+        }
+
+        return $cellAddress1 . ($cellAddress2 ? ":$cellAddress2" : '');
+    }
+
+    private static function adjustSheetTitle(string &$sheetTitle, ?string $value): void
+    {
+        if ($sheetTitle) {
+            $sheetTitle .= '!';
+            if (stripos($value ?? '', $sheetTitle) === 0) {
+                $sheetTitle = '';
+            }
+        }
+    }
+
+    private static function useThisName(string $cellAddress1, string $definedName, ?Worksheet $scope, Worksheet $sheet, string $sheetTitle, ?string $sheetName): bool
+    {
+        return strcasecmp($cellAddress1, $definedName) === 0 && ($scope === null || $scope === $sheet || $sheetTitle === $sheetName);
+    }
+
+    public static function extractCellAddresses(string $cellAddress, bool $a1, Spreadsheet $spreadsheet, Worksheet $sheet, ?string $sheetName = null): array
+    {
+        $cellAddress1 = $cellAddress;
+        $cellAddress2 = null;
+        $namedRanges = $spreadsheet->getNamedRanges();
+        foreach ($namedRanges as $namedRange) {
```

Comment:
The `resolveName()` method in `DefinedName` should handle a lot of this logic for you without needing a loop, including handling the scope checking... it returns either a null, or a `DefinedName` object

Reply:
This isn't possible without a change to resolveName, which I have now added to the PR. The specific case that resolveName had not been able to handle was a reference of the form sheetName!localName, where unqualified localName is out of scope for the referring cell. Excel permits this usage. Adding an optional sheetName parameter to resolveName does indeed simplify the code in Helpers, and doesn't make resolveName too complex.